electrumsv/network_support/mapi.py

```diff
@@ -218,44 +227,56 @@ def validate_json_envelope(json_response: JSONEnvelope) -> None:
+        merkle_proof: bool=False, ds_check: bool=False) -> BroadcastResponse:
+    server.api_key_state[credential_id].record_attempt()
 
     url = server.url if server.url.endswith("/") else server.url +"/"
     url += "tx"
     # It is unclear if we need to pass false values for these, the specification implies that
     # we should but in theory it won't let us broadcast at all.
     params = {
-        "merkleProof": "false",
-        "dsCheck": "false",
+        'merkleProof': 'false' if not merkle_proof else 'true',
+        'merkleFormat': "TSC",
+        'dsCheck': 'false' if not ds_check else 'true',
     }
-    headers = {}
+    if peer_channel is not None:
+        params.update({
+            'callbackURL': peer_channel.get_callback_url(),
+            'callbackToken': f"Bearer {peer_channel.get_write_token().api_key}",
+            # 'callbackEncryption': None  # Todo: add libsodium encryption
+        })
+    headers = {"Content-Type": "application/octet-stream"}
     headers.update(server.get_authorization_headers(credential_id))
     is_ssl = url.startswith("https")
-
     async with aiohttp.ClientSession() as client:
         async with client.post(url, ssl=is_ssl, headers=headers, params=params,
                 data=tx.to_bytes()) as response:
             try:
                 json_response = await decode_response_body(response)
             except (ClientConnectorError, ConnectionError, OSError, SOCKSError):
                 logger.error("failed connecting to %s", url)
+                raise
             else:
                 if response.status != 200:
-                    logger.error("feeQuote request to %s failed with: status: %s, reason: %s",
+                    logger.error("Broadcast request to %s failed with: status: %s, reason: %s",
                         url, response.status, response.reason)
+                    response.raise_for_status()
                 else:
                     assert json_response['encoding'].lower() == 'utf-8'
 
-                    broadcast_response = cast(JSONEnvelope, json_response)
-                    validate_json_envelope(broadcast_response)
+                    broadcast_response_envelope = cast(JSONEnvelope, json_response)
+                    validate_json_envelope(broadcast_response_envelope)
                     logger.debug("transaction broadcast via MAPI server: %s", server.url)
 
                     # TODO(MAPI) Work out if we should be processing the response.
                     # TODO(MAPI) Work out if we should be storing the response.
-                    server_state.record_success()
+                    server.api_key_state[credential_id].record_success()
+                    broadcast_response: BroadcastResponse = \
+                        json.loads(broadcast_response_envelope['payload'])
+                    return broadcast_response
+        raise BroadcastFailedException(f"Broadcast failed for url: {url}, tx: {tx.to_hex()}")
```

Comment:
Is there any path where this is reached? Also, do not put hex tx in strings unless it is temporary or debug logging with deferred formatting where the hex conversion is also deferred in some way.